Feature/dotnet test by project

brandongregoryscott opened the pull request:
Implements feature for issue #52 which allows users to run all test suites in the solution one-by-one, to force serial execution.

Additional notes:

- Leveraged `spawnSync` from the `child_process` module vs. standard `spawn`, since we want to enforce the test runner to run each project in isolation instead of in parallel.
    - This method has a different signature which requires breaking out `dotnet` from each following argument or flag such as `test`, `--no-build`, etc. 
- Updated the original `runBySolution` method to include the coverage flags in the output message so we know 100% what command is being run by the CLI for any debugging purposes vs. running the `dotnet` command directly.
- Added a ternary check for the `signal` arg in the spawn `exit` callback which was generally coming back null, which will default to the status code in the output message
- Although originally noted in the issue, I did not write tests for the sake of time and due to CI not being setup yet. We can create an additional issue to back-fill tests once TravisCI is properly integrated.

wintondeshong commented on cli-dotnet-test.js:
```diff
@@ -22,15 +25,68 @@ const { spawn }   = require("child_process");
+        // leverage the base dotnet test command string here. We'll build out the arg list in an array.
+        let cmd = "dotnet";
+
+        testProjects.map((project) => {
```
Believe this should be a `forEach`, not `map`

wintondeshong replied:
If I'm reading this correctly, this will fail the overall process and short-curcuit once one fails. If that is the case, we should change it to collect the results from each test project and then exit. If one or more failed, exit failed and output a summary of each failed project.

brandongregoryscott replied:
That's a good point, I didn't consider that. Do you think we'd want the ability to toggle a short-circuit or just always have it run the full suite before determining the failure?

wintondeshong replied:
While there could be a scenario where someone would want to exit upon an immediate failure, I've never found it. If its a quick add, go for it, but the default behavior should be to run them all and collect all success and failures before exiting.

wintondeshong commented on cli-dotnet-test.js:
```diff
@@ -85,11 +143,18 @@ const dotnetTest = {
     .usage("option")
     .description(dotnetTest.description())
-    .option("-s, --skip-clean", dotnetTest.descriptionSkipClean())
+    .option("--by-project", "Runs all test projects for the solution in serial")
```
Minor, but should this read as `serially` instead of `in serial` ?

brandongregoryscott replied:
I think you're _technically_ right since serially is an adverb which would describe _Runs_ whereas serial itself is an adjective. I can update that 😛

wintondeshong replied:
One reason why I prefer formal languages 😆

brandongregoryscott commented:
@wintondeshong This should be ready for re-review. I added a `--ci` flag and modified the output to be streaming/captured depending on which mode you are in as we discussed the other day. Plus, some minor refactoring to pull the big callback into its own method and utilize `map()` instead.
